run_flax_speech_recognition_seq2seq_streaming_review.py

```diff
@@ -0,0 +1,1391 @@
+        logger.info(
+            f"  Will skip the first {training_state['step']} steps. If this takes a lot of time,"
+            " you can add the `--ignore_data_skip` flag to your launch command, but you will resume the"
+            " training on data already seen by your model."
+        )
+        for step in tqdm(range(training_state["step"]), desc=f"Skipping data for {training_state['step']} steps...", position=1, leave=False):
+            try:
+                samples = next(train_loader)
+            except StopIteration:
+                epoch += 1
+                train_dataset.set_epoch(epoch)
+                train_loader = data_loader(train_dataset, train_batch_size // num_of_hosts, num_workers=num_workers)
+                samples = next(train_loader)
+            batch = data_collator(samples)
+            # batch = shard(batch.data)
+    
+    for step in tqdm(range(data_args.num_train_steps), desc="Training...", position=1, leave=False):
+        # Skip initial steps if these are specified. 
+        if step < training_state["step"]:
+            continue
+        
+        # =========================== Training ===========================
+        if training_args.do_train:
+            try:
+                samples = next(train_loader)
+            except StopIteration:
+                epoch += 1
+                train_dataset.set_epoch(epoch)
+                train_loader = data_loader(train_dataset, train_batch_size // num_of_hosts, num_workers=num_workers)
+                samples = next(train_loader)
+                logger.info(
+                    f"Completed epoch ({epoch} | Loss: {train_metric['loss']}, Learning Rate:"
+                    f" {train_metric['learning_rate']})"
+                )
+                training_summary["hyperparameters"]["steps_per_epoch"] = step // epoch
+
+            batch = data_collator(samples)
+            batch = shard(batch.data)
+                      
+            state, train_metric = p_train_step(state, batch)
+            train_metrics.append(train_metric)
+
+            train_time += time.time() - train_start
+            train_metric = unreplicate(train_metric)
+
+        # ========================== Evaluating ==========================
+        # Evaluate at each eval_steps, and at the end of training at num_train_steps
+        if training_args.do_eval and (step % training_args.eval_steps == 0 or step == data_args.num_train_steps - 1):
+            logger.info(
+                f"Starting evaluation at step {step} of num_training_step {data_args.num_train_steps} steps. Planned evaluation every {training_args.eval_steps} steps." 
+            )
+            eval_metrics = []
+            eval_preds = []
+            eval_labels = []
+            eval_loader = data_loader(eval_dataset, eval_batch_size, drop_last=False)
+            if data_args.max_eval_samples:
+                max_eval_steps_iter = range(1 + data_args.max_eval_samples // eval_batch_size)
```

Comment:
We can set our max eval samples using `.take`: https://github.com/huggingface/community-events/blob/2eb7fcdad306b5e0d65f10560e259fe398f56723/whisper-fine-tuning-event/run_speech_recognition_seq2seq_streaming.py#L457-L469

Reply:
Then you can just iterate directly over the eval loader

Reply:
We've added the `.take()` / `.select()` now. The problem is when you don't pass `max_eval_samples` but still want evaluation to be done on the entire validation set. In streaming, `len(vectorized_datasets["eval"])` won't work and the only way is to iterate forever until a `StopIteration` exception is raised.